### ISSUE TITLE
'Manually defined' multiplicities are nog recognized as such

### ISSUE THREAD

**Issue by RieksJ** [AmpersandTarski/Ampersand#398]
There should not be a difference in consequences between `r :: A * B [UNI]` and 'r :: A * B     RULE: r~;r |- I'. Currently, however, the prototype (and I guess therefore functional specifications as well) do know the difference. For example, when `r` is used in a prototype INTERFACE, the first variant (with `UNI`) makes that you cannot add multiple B's, whereas the second variant (with the 'manual' univalence rule) produces interfaces that does

**Comment by Michiel-s** [comment]
I'm do not agree with you. As we all know, in Ampersand we do:
1) Conceptual model
2) Rules
3) Interfacing

`r :: A * B [UNI]` specifies the conceptual/data model. This also influences e.g. the database structure. The rules, like `r~;r |- I`, specify additional constraints for this conceptual model. Rules can be enforced as invariant, or process rule.

In my opinion, the facts that you can specify an UNI constraint as rule should not influence the data model. E.g., if you want to specify an UNI constraint as process rule, you can. Without the distinction you cannot have UNI violations (because they are enforced by the database structure) and the process rule is meaningless ór we cannot optimize UNI invariant constraints. Therefore, rules should not influence the interface edit functionality.

**Comment by hanjoosten** [comment]
I fully agree with @RieksJ . The properties defined as part of a relation are merely syntactic suger for a rule. Therefor, no distinction shoud be observable. Hence, this is truly a bug.

**Comment by hanjoosten** [comment]
# Diagnosis
The calculation is done wrong, The function [totsurs](https://github.com/AmpersandTarski/Ampersand/blob/master/src/Database/Design/Ampersand/FSpec/ToFSpec/ADL2FSpec.hs#L220) does not return an expression for this case. 

# Solution 
@stefjoosten , please correct totsurs for this case.

**Comment by hanjoosten** [comment]
A testscript:

```
CONTEXT Bug398 IN ENGLISH
  rProp   :: A * B [UNI] 
  rNoProp :: A * B
  RULE isUni : rNoProp~;rNoProp |- I
ENDCONTEXT
```

**Comment by stefjoosten** [comment]
# Problem

    RELATION r[A*B] [UNI]

    ROLE X MAINTAINS UNI_r

In this situation, the database enforces UNI, and therefore the enforcement by role X is void.

**Comment by RieksJ** [comment]
I consider both arguments of @hanjoosten/me and @Michiel-s to be valid, albeit inconsistent. We cannot resolve the inconsistency in the short run: if we 'fix' this issue as if it were a bug, then Michiel's use-case will not work, introducing practical problems. If we don't do anything about this, then we have theoretical problems. So for now, I guess the way to go is to leave the current implementation alone (to avoid practical problems) and to start thinking about how we can deal with the theoretical ones.

**Comment by RieksJ** [comment]
As far as I can tell, there are several concerns:

1. we (i.e.: at least @hanjoosten and myself) would like the algrebraic notations to be consistent, which means that multiplicities can either be expressed as relation property, e.g. `r :: A * B [UNI]`, and can also be expressed as a rule, e.g. `r~;r |- I`. 
2. we (i.e.: at least @Michiel-s and myself) would like the database to accommodate all transactions that are enabled by `INTERFACE`s, which means that multiplicities such as UNI may be violated while the transaction has not yet been committed.3
3. I would like to see the above two concerns addressed in the presence of one 'belly-context' and multiple 'interface-contexts'. This relates to a text I wrote on contexts, which used to be somewhere on github or the tarski-wiki, but I can no longer find it any more. 
4. I would like to see the above three concerns addressed in a dynamic fashion, i.e. in the advent where users modify the conceptual model in one of the participating contexts, WITHOUT giving up execution speed.

**Comment by hanjoosten** [comment]
2) I would say: **ALL** invariant rules may be violated as long as the transaction has not been committed. (no differentiation between multiplicities and other rules)

**Comment by RieksJ** [comment]
The problem can be visualized as follows. Our current way of thinking is based on the existence of a CONTEXT that contains a valid population, i.e. atoms and pairs that satisfy all invariant rules, and on a database construction algorithm that assumes that the invariants always hold. Since we have come to think about (information)transactions (i.e. cf. the Jan Dietz DEMO model, which are distinct from database transactions), we have run into trouble because we are now obliged to think about populations that are not always within the constraints of the invariants. So either we need a method for doing the computations that are necessary while transactions are not committed (i.e. evaluated various conjuncts), or we think about another algorithm for database construction that allows us to use it in the way we currently do.

My current way of thinking (cf. the not-yet-referable text I wrote on belly-contexts and interface-contexts) is that we

1. explicitly define so-called 'belly-contexts' in terms of concepts, relations and rules (both invariants and (ExecEngine) process rules) as we are used to; such contexts do not specify INTERFACEs.
2. explicitly define so-called 'interface-contexts' in terms of concepts, relations, rules (both invariants and process rules) and INTERFACEs, where the concepts and relations are a subset of those defined by the belly-context to which the interface-context connects (the labels in the INTERFACEs can be seen as belonging to context to which the belly-context interfaces by means of the interface-context). 

Note that the rules in the INTERFACE-contexts, which include multiplicity-constraints on relations that are shared between the contexts, may differ in both contexts. The generator might then be able to construe a database specification that is useable in all contexts, based on the most relaxed (multiplicity) constraints.

Anyone else have other ideas?

**Comment by Michiel-s** [comment]
@hanjoosten how do you want to violate an UNI invariant (specified as UNI) in our current implementation of the database. Even within transactions that is not possible!!

Luckily `r~;r |-` is in the current implementation not syntactic sugar for UNI, so if you specify it as a rule you can allow violations within a transaction.

I agree with Rieks. For now let's leave it as is. And look into it in more detail when we are ready to implement interface specific rule sets.

**Comment by hanjoosten** [comment]
@Michiel-s : Currently, every property of a relation is interpreted as a rule. This is observable with the following script:

```
CONTEXT UNIInvariant IN ENGLISH
   RELATION r[A*B][UNI] = [ ("a1","b1"), ("a1","b2")]
ENDCONTEXT
```
When we try to build a prototype, there is an error message, refering to the rule:

```
Checking on rule violations...
Violations of rule "UNI r[A*B]":
- [("b1","b2"), ("b2","b1")]

Signals for initial population:
Rule(s): ["UNI r[A*B]"]
  Conjunct   : -(r~;r) \/ I[B]
  Violations : [("b1","b2"), ("b2","b1")]

ERROR: No prototype generated because of rule violations.
```

As you can see, the rule is named. We can use this name to refer to it to change the enforcement. (the rule stays the same, however the enforcement goes from invariant to a signal rule:

```
CONTEXT UNISignaled IN ENGLISH
   RELATION r[A*B][UNI] = [ ("a1","b1"), ("a1","b2")]
   ROLE UniSolver MAINTAINS "UNI r[A*B]"
ENDCONTEXT
```

There is nothing wrong with this: The UNI property is exactly the same as the rule. However, there is a bug in the current implementation. That can be demonstrated by trying to build a prototype again:

```
ampersand.exe: !fatal, called at src\Database\Design\Ampersand\FSpec\ToFSpec\ADL2FSpec.hs:547:30 in main:Database.Design.Ampersand.FSpec.ToFSpec.ADL2FSpec
             Ampersand v3.5.2[development:48a98de]
             error nr: 428

```

This error is caused, because when constructing the database, no difference is made between signal rules and invariants. That is wrong.

**Comment by WolframKahl** [comment]
On Fri, Jun 10, 2016 at 07:18:45AM -0700, Han Joosten wrote:
> 2) I would say: **ALL** invariant rules may be violated as long as the transaction has not been committed. (no differentiation between multiplicities and other rules)

Remember that at some point, we will likely have the situation that

  RELATION R[A*Float] [UNI]

will be allowed and enforced by the database,
while writing the corresponding rule will be forbidden,
since it is not checkable by the database.

**Comment by RieksJ** [comment]
Thanks @WolframKahl for this new example of what we can expect to be(come) problematic unless we start to properly make explicit 

1. what the consequences are of what we write in RA; 
2. what we want the database to enforce;
3. how these two things are knitted together;
4. what constraints we apply in terms of contexts, model changes and temporal stuff.

@hanjoosten (and @Michiel-s) already stated the current situation, and there is some dissatisfaction about this current state. Perhaps we should first clarify how things work today, and then create a vision of how we see that things might evolve, and create a roadmap from there - just to make things manageable.

**Comment by stefjoosten** [comment]
Good point, Wolfram. `FLOAT`s come with a more fundamental issue, though. The entire Ampersand world is built on the assumption that equality is defined for atoms. `FLOAT` is of course the most notable theoretical exception, at least of we want them to behave (mathematically correct) like real numbers. The question is whether we want to loosen Ampersand's equality assumption to include real numbers into the language. I would say this is worth another discussion. And practically, this affects so much in the entire implementation of Ampersand that we need to consider the wisdom of changing Ampersand to this respect. It might even be wiser to build a different language for that purpose.

As for `FLOAT`s, I think we may adopt one or more of the practical ways of treating `FLOAT`s, as seen in other programming languages. For now I would say we either use a `FLOAT` with equality (which would be binary equality) or refrain from `FLOAT`s entirely.

**Comment by hanjoosten** [comment]
The 'other discussion' is the one mentioned in issue #341

**Comment by WolframKahl** [comment]
The recent posts made me remember a different issue I wanted to mention here: With
```
RELATION R[A*B] [UNI,INJ]
```
, can the database be made to enforce both `UNI` and `INJ`? I seem to recall from a different issue that conceptally, it might be considered unspecified which of the tables `A` and `B` the new relation `R` goes into, therewith having the database enforcing only one of `UNI` and `INJ`. But if the interface then allows the other one to be temporarily violated, that would mean that the choice between `A` and `B` becomes user-observable, which in my eyes is an argument against leaving it unspecified.

A follow-up question would then be whether you want to keep it simple (e.g. give `UNI` precedence over `INJ`), or document all the cases to allow the better choice of `INJ` for something with `[UNI,INJ,SURJ]`.

Or, alternatively, one could make the front-end explain the backend's choice sufficiently clearly.

In any case, I don't quite understand why it is so desirable to be able to temporarily violate constraints/rules.

**Comment by hanjoosten** [comment]
It hasn't been said that there is any user-observable behaviour. The user doesn't 'see' the database, she only sees the user interface. which is generally a different view/context.The transaction is done in several phases (See [DEMO model](https://en.wikipedia.org/wiki/Design_%26_Engineering_Methodology_for_Organizations)) In Ampersand, we do not take this into account well enough, which is probably one of the causes for the problems we are facing. 

The last question is simple: While it is obvious we want the persistent storage to hold only data sets that are compliant to the invariant rules, during a transaction it is also obvious that rules are required to be violated. As an axample, lets say that each house must have one owner:

```
RELATION owner[House*Person][UNI,TOT]
```
In that database, it is perfectly allright to have a mandatory attribute holding the Id of a person in the table `house`.

Of course, sometimes homes are sold, which leads to a transfer in ownership. In the user interface, we have a screen, with a field where we want to erase the `old` owner, and fill in the `new` owner. In the user interface, this cannot allwasy be an atomic action. (maybe this one could be made so, but if things get slightly more complicated, the user needs to temporarily violate rules).

**Comment by WolframKahl** [comment]
> with a field where we want to erase the old owner, and fill in the new owner. In the user interface, this cannot allwasy be an atomic action. (maybe this one could be made so

I think it should be made so ─ isn't this what the “U” of CRUD is about? I don't think an update should necessarily be modelled/implemented as a sequence of delete-create...

**Comment by hanjoosten** [comment]
As I mentioned, there could be more complex cases. Consider some concept that is source of multiple relations, that are all TOT. How would the user interface look like, if a new instance of that concept was to be added, without temporarily be able to violate any rules?

**Comment by WolframKahl** [comment]
On Tue, Jun 14, 2016 at 06:47:31AM -0700, Han Joosten wrote:
> As I mentioned, there could be more complex cases. Consider some concept that is source of multiple relations, that are all TOT. How would the user interface look like, if a new instance of that concept was to be added, without temporarily be able to violate any rules?

Certainly. However, I still think that one should be grateful for any constraint that can be enforced by the system so that it cannot be broken even temporarily, as long as all desired changes can somehow be achieved.

That is, concentrate on how to achieve desired changes.

And don't think about how to break invariants temporarily, because that as such has no value.

**Comment by stefjoosten** [comment]
The same issue popped up just now in issue  #728.

**Comment by Michiel-s** [comment]
I propose to close this issue as there is no issue at all.

To summarize:
1. Writing `relation [UNI]` is syntactical sugar for the rule `r~;r |- I` as far as it concerns the rule evaluation
2. However, for relation definitions we also keep track of the defined properties (see Relation data type in AbstractSyntaxTree.hs). The properties UNI and INJ are included in the decision to add the relation registration to the concept table of the SRC/TGT or make it a binary table.
3. Therefore, when writing the multiplicity constraint by means of [UNI] or [INJ] it cannot be violated in the prototypes, because there is simply no possibility to hold 2 values during the transaction. Keep in mind that we use the database (and generated SQL queries) to evaluate our rules runtime.
4. Design time this is different and there is indeed no difference (because there is no persistent storage).

So if you want to allow for violations of the constraint and/or let the rule be maintained by a role, write them as rules... if you don't care and just want to specify an invariant rule and see that reflected in the data model, use relation property syntax.
